_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+                        consumer.commitAsync(pendingOffsets, this);
+                        pendingOffsets.clear();
+                    }
+                } catch (OffsetOutOfRangeException | NoOffsetForPartitionException e) {
+                    // Handle offset errors
+                    System.out.println("Invalid or no offset found, using latest");
+                    consumer.seekToEnd(e.partitions());
+                    consumer.commitSync();
+                } catch (Exception e) {
+                    // Handle other exceptions, including retriable ones
+                    System.err.println(e.getMessage());
+                    if (!retriable(e)) {
+                        e.printStackTrace();
+                        System.exit(1);
+                    }
+                }
+            }
+        }
+    }
+
+    private KafkaConsumer<Long, byte[]> createKafkaConsumer() {
+        // Create properties for the Kafka consumer
+        Properties props = new Properties();
+
+        // Configure the connection to Kafka brokers
+        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BOOTSTRAP_SERVERS);
+
+        // Set a unique client ID for tracking
+        props.put(ConsumerConfig.CLIENT_ID_CONFIG, "client-" + UUID.randomUUID());
+
+        //Set a consumer group ID for the consumer
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
+
+         // Configure deserializers for keys and values
+        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, LongDeserializer.class);
+        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
+
+        // Disable automatic offset committing
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
+
+        //Set the offset reset behavior to start consuming from the earliest available offset
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        return new KafkaConsumer<>(props);
+    }
+
+    private void sleep(long ms) {
+        try {
+            TimeUnit.MILLISECONDS.sleep(ms);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean retriable(Exception e) {
+        if (e == null) {
+            return false;
+        } else if (e instanceof IllegalArgumentException
+            || e instanceof UnsupportedOperationException
+            || !(e instanceof RebalanceInProgressException)
+            || !(e instanceof RetriableException)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+        System.out.printf("Assigned partitions: %s%n", partitions);
+    }
+
+    @Override
+    public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        System.out.printf("Revoked partitions: %s%n", partitions);
+        kafkaConsumer.commitSync(pendingOffsets);
```

Comment:
Same here, just `kafkaConsumer.commitSync();` should be enough.

Reply:
@fvaleri - and here?

Reply:
@showuon I think kafkaConsumer.commitSync() commits the offsets of all records taken with the current poll. Instead, here we only want to commit processed records. This is why I'm using pending offsets in the first place.

Reply:
OK, let's keep it.